plugins/modem-manager/fu-plugin-modem-manager.c

```diff
@@ -0,0 +1,472 @@
+	fu_plugin_cache_remove (plugin, priv->inhibited->physical_id);
+	fu_plugin_device_remove (plugin, FU_DEVICE (dev));
+
+	/* no need to wait for more ports, cancel that right away */
+	if (priv->inhibited->udev_timeout_id != 0) {
+		g_source_remove (priv->inhibited->udev_timeout_id);
+		priv->inhibited->udev_timeout_id = 0;
+	}
+}
+
+static void
+fu_plugin_mm_uninhibit_device (FuPlugin *plugin)
+{
+	FuPluginData *priv = fu_plugin_get_data (plugin);
+	g_autoptr(FuPluginMmInhibitedDevice) info = NULL;
+
+	/* get the device removed from the plugin cache before uninhibiting */
+	fu_plugin_mm_udev_device_removed (plugin);
+
+	info = g_steal_pointer (&priv->inhibited);
+	if ((priv->manager != NULL) && (info != NULL)) {
+		g_debug ("uninhibit modemmanager device with uid %s", info->inhibited_uid);
+		mm_manager_uninhibit_device_sync (priv->manager, info->inhibited_uid, NULL, NULL);
+	}
+}
+
+static gboolean
+fu_plugin_mm_udev_device_ports_timeout (gpointer user_data)
+{
+	FuPlugin *plugin = user_data;
+	FuPluginData *priv = fu_plugin_get_data (plugin);
+	FuMmDevice *dev;
+	g_autoptr(GError) error = NULL;
+
+	g_assert (priv->inhibited != NULL);
+	priv->inhibited->udev_timeout_id = 0;
+
+	dev = fu_plugin_cache_lookup (plugin, priv->inhibited->physical_id);
+
+	if (dev != NULL) {
+		if (!fu_device_probe (FU_DEVICE (dev), &error)) {
+			g_debug ("failed to probe MM device: %s", error->message);
+		} else {
+			fu_plugin_device_add (plugin, FU_DEVICE (dev));
+		}
+	}
+
+	return G_SOURCE_REMOVE;
+}
+
+static void
+fu_plugin_mm_udev_device_ports_timeout_reset (FuPlugin *plugin)
+{
+	FuPluginData *priv = fu_plugin_get_data (plugin);
+
+	g_assert (priv->inhibited != NULL);
+	if (priv->inhibited->udev_timeout_id != 0)
+		g_source_remove (priv->inhibited->udev_timeout_id);
+	priv->inhibited->udev_timeout_id = g_timeout_add_seconds (FU_MM_UDEV_DEVICE_PORTS_TIMEOUT,
+								  fu_plugin_mm_udev_device_ports_timeout,
+								  plugin);
+}
+
+static void
+fu_plugin_mm_udev_device_port_added (FuPlugin		*plugin,
+				     const gchar	*subsystem,
+				     const gchar	*path,
+				     gint		 ifnum)
+{
+	FuPluginData *priv = fu_plugin_get_data (plugin);
+	FuMmDevice *existing;
+	g_autoptr(FuMmDevice) dev = NULL;
+	g_autoptr(GError) error = NULL;
+
+	g_assert (priv->inhibited != NULL);
+	existing = fu_plugin_cache_lookup (plugin, priv->inhibited->physical_id);
+	if (existing != NULL) {
+		/* add port to existing device */
+		fu_mm_device_udev_add_port (existing, subsystem, path, ifnum);
+		fu_plugin_mm_udev_device_ports_timeout_reset (plugin);
+		return;
+	}
+	/* create device and add to cache */
+	dev = fu_mm_device_udev_new (priv->manager,
+				     priv->inhibited->physical_id,
+				     priv->inhibited->vendor,
+				     priv->inhibited->name,
+				     priv->inhibited->version,
+				     (const gchar **) priv->inhibited->guids,
+				     priv->inhibited->update_methods,
+				     priv->inhibited->detach_fastboot_at,
+				     priv->inhibited->port_at_ifnum);
```

Comment:
Would just passing `priv->inhibited` be simpler?

Reply:
Yes, possibly. I'll refactor it a bit to have the FuPluginMmInhibitedDevice implementation in its own set of .h/.c files.

Reply:
I moved the struct and wrote a new/free pair of functions for it; thought of making it a GObject with per-field getters but I don't think that made much sense.